cpp/src/url.cpp

```diff
@@ -274,14 +274,10 @@ std::string to_string(const url& u) {
 std::istream& operator>>(std::istream& i, url& u) {
     std::string s;
     i >> s;
-    if (!i.fail() && !i.bad()) {
-        if (!s.empty()) {
-            url::impl* p = new url::impl(s);
-            p->defaults();
-            u.impl_.reset(p);
-        } else {
-            i.clear(std::ios::failbit);
-        }
+    if (!i.fail()) {
```

Comment:
```suggestion
    if (i) {
```

I think that this makes the most sense. It has the same meaning as `!i.fail() && !i.bad()`, according to https://www.cplusplus.com/reference/ios/ios/operator_bool/ and it is shorter. But tbh, I never really worked with C++ streams, so I may be missing something.

Reply:
I think that the premise that you can't get s.empty by itself might be wrong - it looks to me that getting end-of-file is independent from fail and bad so I think it should be possible to eof and s.empty() but !fail && !bad. However I'm not any expert on iostreams either and that's just my reading of a manpage or two.

Reply:
> It has the same meaning as `!i.fail() && !i.bad()`

It does, but never mind that. `fail()` also checks the badbit, as shown in the colorful table at https://en.cppreference.com/w/cpp/io/basic_ios/fail.so what you have there now is probably best.

>  it should be possible to eof and s.empty() but !fail && !bad

@DreamPearl could you investigate this more? I myself think it is not possible, but we need to be sure about this, before we make the change.

Reply:
@astitcher @jiridanek I tried to investigate more into it and that leads me to think that `end-of-file sets the failbit.`
I tried to run this sample code: http://cpp.sh/3gsgh

And below are some references. Please let me know what do you think of this?

`“If no characters are extracted then std::ios::failbit is set.”`
Link: https://en.cppreference.com/w/cpp/string/basic_string/operator_ltltgtgt

`“Operations that attempt to read at the End-of-File fail, and thus both the eofbit and the failbit end up set. This function can be used to check whether the failure is due to reaching the End-of-File or to some other reason.”`
LInk: http://www.cplusplus.com/reference/ios/ios/eof/

`“Reaching the End-of-File sets the eofbit. But note that operations that reach the End-of-File may also set the failbit if this makes them fail (thus setting both eofbit and failbit).”`
Link: http://www.cplusplus.com/reference/ios/ios/fail/

Reply:
@DreamPearl you've convinced me! Thanks for the thorough work.

I also prefer the ```i.fail()``` version

Reply:
Thanks!